### ISSUE TITLE
Keras model serving for inference with dl4j

### ISSUE THREAD

**Issue by DreamyKten** [deeplearning4j/deeplearning4j#8298]
> 1. I saved a keras model with weights with python and loaded a `ComputationGraph` in java on DL4j using

`model = KerasModelImport.importKerasModelAndWeights(unet, enforceTrainingConfig=false)`

> 2. I create my input `INArray` with `Nd4j.create(floats)`
and run the inference : 

`INDArray output = model.output(input)`

> 3. I retrieve the output : 

 `float[][] x = output.reshape(new int[]{floats.length, 256 * 256}).toFloatMatrix();`

<b> There is a output but similar to the input with some different pattern. The result is not correct. </b> 

I tried this model, it works well in python with Keras and in Java with a frozenModel (freezegraph) on Tensorflow. 
Does anyone try to use trained keras model on dl4j ?   

I expect to reproduce the inference on dl4j as I got on Tensorflow. 

Have you ever encountered this issue ? any hints will be helpful, thanks.
https://stackoverflow.com/questions/58434187/keras-model-serving-for-inference-with-dl4j

**Comment by DreamyKten** [comment]
Seem to be the same issue : https://github.com/eclipse/deeplearning4j/issues/8295

input : https://i.stack.imgur.com/n8DQ5.png

output : https://i.stack.imgur.com/IUHuj.png

**Comment by guigautier** [comment]
Seem to be the same issue : https://github.com/eclipse/deeplearning4j/issues/8295

**Comment by eraly** [comment]
Are you using beta4? There is a bug that was fixed in beta5.

**Comment by DreamyKten** [comment]
I'm using beta5 

This is my maven dependencies : 

       <dependency>
            <groupId>org.nd4j</groupId>
            <artifactId>nd4j-api</artifactId>
            <version>1.0.0-beta5</version>
        </dependency>

        <dependency>
            <groupId>org.deeplearning4j</groupId>
            <artifactId>deeplearning4j-core</artifactId>
            <version>1.0.0-beta5</version>
        </dependency>

        <dependency>
            <groupId>org.deeplearning4j</groupId>
            <artifactId>deeplearning4j-nn</artifactId>
            <version>1.0.0-beta5</version>
        </dependency>

        <dependency>
            <groupId>org.deeplearning4j</groupId>
            <artifactId>deeplearning4j-modelimport</artifactId>
            <version>1.0.0-beta5</version>
        </dependency>
        <dependency>
            <groupId>org.deeplearning4j</groupId>
            <artifactId>deeplearning4j-cuda-10.0</artifactId>
            <version>1.0.0-beta5</version>
        </dependency>
        <dependency>
            <groupId>org.deeplearning4j</groupId>
            <artifactId>deeplearning4j-parallel-wrapper</artifactId>
            <version>1.0.0-beta5</version>
        </dependency>

**Comment by eraly** [comment]
Thank you for the update. Will you please upload either your model or the model architecture (json)? Also, what OS and what version of Keras?

**Comment by DreamyKten** [comment]
I'm using Windows 10 and Keras 2.2.4 with tensorflow  backend 1.13.1. 
And this is my model 
[unet.zip](https://github.com/eclipse/deeplearning4j/files/3761049/unet.zip)

Thank you.

**Comment by AlexDBlack** [comment]
Here's my attempt to reproduce this from the provided JSON configuration.
I basically overfit the provided image, then compared results in DL4J vs. Keras.

I am seeing a difference...
Absolute difference over all output pixels:
```
Min diff: 1.1920928955078125E-7
Max diff: 0.4378690719604492
Avg diff: 0.027073459699749947
```

![image](https://user-images.githubusercontent.com/2360237/67390139-c1d93980-f5e7-11e9-90fd-135ca233c9d0.png)

So, not as extreme as your images, but still slightly "washed out" and corner artifact...

Code to reproduce:
https://gist.github.com/AlexDBlack/3a4f58edcf243ef4d3faa73ee176eb04

So, I'd say this is confirmed as a bug of some description. Will look into it further and try to isolate it.

**Comment by AlexDBlack** [comment]
So, it's definitely coming from deconv layers. If I use the same input image as the keras model in DL4J, activation differences are around 1e-8 on average for the conv layers... until the first deconv (conv2d_transpose_1)
Code and output: https://gist.github.com/AlexDBlack/abb3bf0a6f4f384863fa467f498447a4
I'm using Keras 2.3.1 btw.

Note difference is a little bigger with NativeImageLoader, not sure why (probably just slightly different resize and/or grayscale conversion algorithms or something).

Will look at this more tomorrow, unless @eraly debugs this first.

**Comment by AlexDBlack** [comment]
So, a few things I've noticed here

First: this padding calculation is wrong, but it's unused anyway, and will be calculated/replaced in the c++ op instead:
https://github.com/eclipse/deeplearning4j/blob/master/deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/convolution/Deconvolution2DLayer.java#L202
(we end up with "padding" calculated as [24, 24] in this case)


Second: bias add appears to be hardcoded to NCHW format:
https://github.com/eclipse/deeplearning4j/blob/master/libnd4j/include/ops/declarable/generic/nn/convo/deconv2d.cpp#L84

I believe this should be `helpers::addBias(block, *output, *bias, *output, isNCHW);`

**Comment by eraly** [comment]
The C++ implementation gives incorrect answers. Here is a simple test case. The expected answers are calculated via Keras. Keras code is included in the gist. Attaching the h5 model from Keras if someone wants to debug the java side. For the c++ side you don't need anything else other than what is in the gist since I manually recreate the call to libnd4j

https://gist.github.com/eraly/49dfdb401b0347d8183027edae462f3e

Keras Model:
[de_conv.h5.zip](https://github.com/eclipse/deeplearning4j/files/3765088/de_conv.h5.zip)

**Comment by eraly** [comment]
For a case with biases not set to zero:
https://gist.github.com/eraly/0824294d453261c58ef0585d70496464

Keras Model:
[de_conv.h5.zip](https://github.com/eclipse/deeplearning4j/files/3765115/de_conv.h5.zip)

**Comment by eraly** [comment]
Also of note: With the case here (i.e combination of kernel, stride and input sizes) the output shape with same == output shape with zero padding and not same. This gives the right answer. So it might be an issue only with same mode?

**Comment by AlexDBlack** [comment]
Update here: looks like our shape and padding calculation may not be right in c++, in particular the args passed to col2im. I'll go through this and work out what it should be...
Pg24-26 here in case anyone wants to know: https://arxiv.org/pdf/1603.07285.pdf

**Comment by raver119** [comment]
addBias being hardcoded to NCHW is fine, since temporary output array is always in NCHW format.

**Comment by AlexDBlack** [comment]
Fixed here: https://github.com/KonduitAI/deeplearning4j/pull/12
I added a whole lot of test cases pulled from Keras here also https://github.com/KonduitAI/dl4j-test-resources/pull/2
2 test cases left there (valid mode, output size differs; separate issue to same mode problem here).

I should be able to merge that tomorrow, then push that back to Eclipse. New snapshots with the fix will likely be up some time next week.

**Comment by AlexDBlack** [comment]
Thanks for reporting this. The fix has been merged here: https://github.com/KonduitAI/deeplearning4j/pull/12
This will be merged back to the Eclipse repo shortly, and will be available on snapshots some time next week.

![image](https://user-images.githubusercontent.com/2360237/67612357-d25b0100-f7ed-11e9-96be-3fdcffeec816.png)

Keras vs. DL4J layer activations comparison from earlier:
https://gist.github.com/AlexDBlack/fc25ead7c88f207fa7ca9f43154014c7
